cpp/src/aztec/honk/pcs/gemini/gemini.hpp

```diff
@@ -298,135 +251,144 @@ template <typename Params> class MultilinearReductionScheme {
         // After the swap, we have
         // tmp = G(X)/r
         // A_0_neg = F(X) (since &batched_F == &A_0_neg)
-        Polynomial& A_0_neg = witness_polynomials[1];
+        Polynomial& A_0_neg = fold_polynomials[1];
 
         // A₀₋(X) = F(X) - G(X)/r, s.t. A₀₋(-r) = A₀(-r)
         A_0_neg -= tmp;
 
         /*
-         * compute rₗ = r^{2ˡ} for l = 0, 1, ..., m-1
+         * Compute the m evaluations a_0 = Fold_{-r}^(0)(-r), and a_l = Fold^(l)(-r^{2^l}), l = 1, ..., m-1.
+         * Add them to the transcript
          */
-        std::vector<Fr> r_squares = squares_of_r(r, num_variables);
-
-        // evaluate all new polynomials Aₗ at -rₗ
-        std::vector<Fr> evals;
-        evals.reserve(num_variables);
+        std::vector<Fr> fold_polynomial_evals;
+        fold_polynomial_evals.reserve(num_variables);
         for (size_t l = 0; l < num_variables; ++l) {
-            const Polynomial& A_l = witness_polynomials[l + 1];
-            const Fr r_l_neg = -r_squares[l];
-            evals.emplace_back(A_l.evaluate(r_l_neg));
-        }
+            const Polynomial& A_l = fold_polynomials[l + 1];
 
-        /*
-         * Add evaluations a_i, i = 0,...,m-1 to transcript
-         */
-        for (size_t i = 0; i < evals.size(); ++i) {
-            std::string label = "a_" + std::to_string(i);
-            transcript->add_element(label, evals[i].to_buffer());
+            fold_polynomial_evals.emplace_back(A_l.evaluate(-r_squares[l]));
+            transcript->add_element("a_" + std::to_string(l), fold_polynomial_evals[l].to_buffer());
         }
 
-        /*
-         * Construct the 'Proof' which consists of:
-         * (1) The m-1 commitments [Fold^{l}], l = 1, ..., m-1
-         * (2) The m evaluations a_0 = Fold_{-r}^(0)(-r), and a_l = Fold^(l)(-r^{2^l}), l = 1, ..., m-1
-         */
-        Proof<Params> proof = { commitments, evals };
+        // Compute evaluation A₀(r)
+        auto a_0_pos = compute_eval_pos(batched_evaluation, mle_opening_point, r_squares, fold_polynomial_evals);
```

Comment:
It may be more efficient to simply evaluate `A_0(r)` directly since memory access for polynomial evaluation is much more sequential. The function `compute_eval_pos` performs several field inversion that can be a quite expensive.

Reply:
Good call - done